src/rai_bench/rai_bench/tool_calling_agent_bench/agent_bench.py

```diff
@@ -0,0 +1,211 @@
+            try:
+                response = agent.invoke(
+                    {"messages": [HumanMultimodalMessage(content=task.get_prompt())]},
+                    config=config,
+                )
+                task.verify_tool_calls(response=response)
+            except GraphRecursionError as e:
+                self.logger.error(f"Graph Recursion Error: {e}")
+                task.result.errors.append(f"Graph Recursion Error: {e}")
+            te = time.perf_counter()
+            total_time = te - ts
+            result = task.result
+            for callback in callbacks:
+                self.score_tracing_handler.send_score(
+                    callback=callback,
+                    run_id=run_id,
+                    success=result.success,
+                    errors=result.errors,
+                )
+
+            self.logger.info(
+                f"TASK SUCCESS: {result.success}, TOTAL TIME: {total_time:.3f}"
+            )
+
+            task_result = TaskResult(
+                task=task.get_prompt(),
+                model=model_name,
+                success=result.success,
+                errors=result.errors if result.errors else [],
+                total_time=total_time,
+                run_id=run_id,
+            )
+            self.task_results.append(task_result)
+
+            if model_name not in self.model_results:
+                self.model_results[model_name] = []
+            self.model_results[model_name].append(task_result)
+
+            self._save_task_result_to_csv(task_result)
+
+            completed_tasks = sum(
+                len(results) for results in self.model_results.values()
+            )
+            if completed_tasks == self.num_tasks:
+                self._compute_and_save_summary()
+
+        except StopIteration:
+            if self.task_results:
+                self._compute_and_save_summary()
+            print("No more scenarios left to run.")
+
+    def _save_task_result_to_csv(self, result: TaskResult) -> None:
+        """Save a single task result to the CSV file."""
+        with open(
+            self.results_filename, mode="a", newline="", encoding="utf-8"
+        ) as file:
+            writer = csv.DictWriter(file, fieldnames=self.fieldnames)
+            writer.writerow(result.model_dump())
+
+    def _initialize_results_file(self):
+        """Initialize the CSV file with headers."""
+        with open(
+            self.results_filename, mode="w", newline="", encoding="utf-8"
+        ) as file:
+            writer = csv.DictWriter(file, fieldnames=self.fieldnames)
+            writer.writeheader()
+
+    def _initialize_summary_file(self):
+        """Initialize the summary CSV file with headers."""
+        with open(
+            self.summary_filename, mode="w", newline="", encoding="utf-8"
+        ) as file:
+            writer = csv.DictWriter(file, fieldnames=self.summary_fieldnames)
+            writer.writeheader()
```

Comment:
How about refactoring to 2 functions/staticmethods:
```python
def csv_initialize(filename, base_model): 
    ...
def csv_writerow(filename, base_model_cls):
```

You would not need to keep `self.fieldnames` and `self.summary_fieldnames`